### PR TITLE
Added new features

### PR DESCRIPTION
**Pull request by mrboring**
Not sure is you are interested, but I've added some new features:

* Window scalable
* Panels resizable
* Added line numbers
* Added start/end line numbers
* Added start/end column numbers
* Smaller font size
* Added FontSize parameter
* Added filename to window title
* Refactored into separate files

### CODE REVIEW

**Comment by lzybkr** [comment]
Thanks. This project was a quick and dirty demo for use in a presentation and people do seem to use it, so enhancements are welcome.

General feedback:

The selected text doesn't match the ast when I tried `Show-Ast $function:prompt` where the function is defined in `profile.ps1`.

The line numbers in the right pane should probably match the original file instead of being 1 based.

Ideally the line numbers in the right pane should not be editable. This might be hard to implement though. A user can press <kbd>F5</kbd> to refresh the ast after editing, and it is pretty confusing with the line numbers even though it seems to work even when you do weird things like delete a line number or leave them.

I like the font size parameter, but do think the default font size is too small, especially given that this module is used for presentations.

The extent in the ast pane is a little confusing, it seems to be `start-line,end-line/start-column,end-column/start-offset,end-offset` but if feels unnecessarily noisy. Maybe I never should have put the offsets there in the first place, but I did use the open set notation to suggest the end offset is not included in the range. At any rate, I think your changes make it even more confusing.

I do like seeing the rest of the line in the ast pane, that is a big improvement.

Resizing is really handy, I recall being annoyed during the presentation that it was I hadn't added that feature,

**Comment by mrboring** [comment]
Thanks for the detailed feedback! 

I've got a few more tasks to complete. Mainly tweaking an tidying up.

I've added commits that address the items below.

> The selected text doesn't match the ast when I tried `Show-Ast $function:prompt` where the function is defined in `profile.ps1`.

This is fixed.

> The line numbers in the right pane should probably match the original file instead of being 1 based.

This is fixed.

> Ideally the line numbers in the right pane should not be editable. This might be hard to implement though. A user can press F5 to refresh the ast after editing, and it is pretty confusing with the line numbers even though it seems to work even when you do weird things like delete a line number or leave them.

I looked at other methods, like having a listview to the left of the scriptview. Populating it with line numbers and trying to sync it with the script view as the user scrolls. I decided that this was too much work for now. I may revisit it later.

The line numbers are removed and re-added on `F5`. So, the user can mess them up and they will be fixed.

> I like the font size parameter, but do think the default font size is too small, especially given that this module is used for presentations.

I’ve hard coded fallback defaults into `Get-ShowPsAstConfig`. The default values are:
```
FontSize          = 12
ExtentDetailLevel = 'Normal'
```

These defaults can be overridden by editing  `<module root>\config.txt`. The default contents:
```
FontSize = 12

# Valid values are: 
# Normal (line numbers only) 
# Detailed (line numbers, column numbers, offset numbers)
ExtentDetailLevel = Detailed
```

If the  `<module root>\config.txt` file gets deleted the values in `Get-ShowPsAstConfig` will be used

Also, they can be overridden by parameters for `Show-Ast`:

```
-FontSize
-ExtentDetailLevel
```

> The extent in the ast pane is a little confusing, it seems to be `start-line,end-line/start-column,end-column/start-offset,end-offset` but if feels unnecessarily noisy. Maybe I never should have put the offsets there in the first place, but I did use the open set notation to suggest the end offset is not included in the range. At any rate, I think your changes make it even more confusing.

Normally line numbers are the most helpful. Before I added them it took a while to find the AST pane item related to a specific line in the scriptview. The other information is mainly useful for debugging issues. 

I ‘ve added a parameter: ExtentDetailLevel. It can have the following values:
```
Normal = line numbers only
Detailed = line numbers, column numbers and offset numbers
```

**Comment by mrboring** [comment]
I've finished by updates. I'm not expecting any more commits unless I find issues.

**Comment by mrboring** [comment]
I found the ordering of items in the treeview was not always consistent. I've added a commit to address this.

**Comment by mrboring** [comment]
I remove `Closes #<issue number>` and `Resolves #<issue number>` from the commit messages. These may have casused issues if the pull request is merged.

**Comment by mrboring** [comment]
Added run without blocking command line.

**Comment by mrboring** [comment]
Removed my personal email address from commits.

File: ShowPSAst.psm1 [hunk 1]
```diff
@@ -1,255 +1,23 @@
...
+Foreach ($import in @($public + $private)) {
+    Try {
+        . $import.fullname
+    }
+    Catch {
+        Write-Error -Message "Failed to import function $($import.fullname): $_"
     }
 }
 
+$showPsAstConfig = Get-ShowPsAstConfig
```
Comment (lzybkr) [inline ShowPSAst.psm1]: I think defaults should be loaded as needed so changes to the configuration that happen after module load are honored.

**Review by lzybkr — CHANGES_REQUESTED** [review]
I finally took a quick look at the code and I'm not going to nitpick anything, overall it looks pretty good, just one comment on the logic regarding defaults.

I'd ask that you change the encoding of ShowPSAst.psd1 so it changes from a binary file to a text file (I think this was an unfortunate choice in `New-ModuleManifest` which I likely used).

**Comment by mrboring** [comment]
> I finally took a quick look at the code and I'm not going to nitpick anything, overall it looks pretty good, just one comment on the logic regarding defaults.

* I've moved the call to `Get-ShowPsAstConfig` inside of `Show-Ast`
* I'm now passing `Show-Ast` parameters into `Get-ShowPsAstConfig`
* `Get-ShowPsAstConfig` will have layered configuration priorities in this order:
    - `Show-Ast` parameters - if provided
    - Coniguration file - if found
    - Hard coded values
* No commit for this yet as I'm still testing it

With regard to nitpicking, please feel free to ask me to change anything you are not happy with.

> I'd ask that you change the encoding of ShowPSAst.psd1 so it changes from a binary file to a text file (I think this was an unfortunate choice in `New-ModuleManifest` which I likely used).

Is converting the encoding from UTF-16 LE to UTF-8 enough to achieve this?

**Comment by lzybkr** [comment]
> Is converting the encoding from UTF-16 LE to UTF-8 enough to achieve this?

UTF8 without a BOM is fine.

**Comment by mrboring** [comment]
I've changed the encoding to `UTF-8`. However, `GitHub`  still thinks the file is binary. 

image

After some searching, I think it my be related to `Git` not knowing the file extension `.psd1`. Take a look at:

[Why does Git treat this text file as a binary file?](https://stackoverflow.com/questions/6855712/why-does-git-treat-this-text-file-as-a-binary-file)

The recommendation is to create a `.gitattributes` file. I've not done this before, but I'm happy to look into it.

**Comment by mrboring** [comment]
After some messing around I think Git/GitHub now sees `ShowPSAst.psd1` as a text file.

**Comment by mrboring** [comment]
I've completed the changes you requested.

**Comment by mrboring** [comment]
`Get-Ast` was being called outside of the runspace and values from the extent were passed into the runspace. Inside the runspace I was calling `Get-Ast` again using the `Extent.Text` property as `InputObject`. Passing the AST into the runspace means that the second call to `Get-Ast` is no longer required.

Background: I did it this way because I was not familiar with creating runspaces. So, the first version I did was based on jobs. With jobs you can only pass in arguments and not parameters. When I passed in the AST as a parameter it was converted to text.

**Review by lzybkr — APPROVED** [review]

**Comment by lzybkr** [comment]
LGTM

I'll squash the commits to merge. If you'd like a specific commit message, go ahead and suggest something, otherwise I'll just take your first comment in the PR.

**Comment by mrboring** [comment]
First comment on PR would be OK.

**Comment by fflaten** [comment]
@lzybkr Any chance of getting a new PSGallery-release? 🙂

**Comment by mklement0** [comment]
@mrboring, this looks like a great addition (still unreleased, as @fflaten observes), thank you, but I've noticed one problematic aspect:

* All source code files now have `Set-StrictMode -Version Latest` at the top.

* However, `Set-StrictMode -Version Latest` should never be used in production code, as it runs the risk of breaking the code should additional strictness checks be introduced in the future that the code doesn't meet. Instead, lock in the specific version that was latest at the time of authoring the code.

  * That said, in practice the risk is perhaps non-existent, given that PowerShell v3 was the last version for which additional checks were introduced.

**Comment by mrboring** [comment]
@mklement0 Thanks for the feedback.

**Pull request merged**